openedx/openedx-platform
# feat: make hide from TOC a visibility section setting

## Description

This PR exposes the `hide_from_toc` xblock attribute so course authors can configure it as a section visibility option in Studio. Before this change, the Hide from TOC functionality was mainly used by OLX components. Hence, it wasn't available for configuration through the Studio UI. Still, its implementation existed in the platform and could be used by setting the attribute: `hide_from_toc=true` as part of the OLX definition. 

So we could use the `hide_from_toc` attribute in the Studio UI making it configurable for sections and their children, we did the following:

- Add `hide_from_toc` to the CMS Xblock handlers, serializers, and data classes so it can be recognized when setting the new visibility state option
-  Add it to the inheritance mixin so the sections' children recognize the configuration option
- Make it available in the backbone frontend for course authors to configure as in `visible_to_staff_only`

## Supporting information

These changes are part of the effort made to implement [feature enhancement proposal: Hide sections from course outline](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3853975595/Feature+Enhancement+Proposal+Hide+Sections+from+course+outline)

## Testing instructions

1. Move into this branch
2. Collect statics by running: `tutor dev exec lms openedx-assets build --env=dev`
3. Enable TOC UI by setting: `FEATURES["ENABLE_HIDE_FROM_TOC_UI"] = True` in your CMS environment, this will allow course staff to configure the hide from toc option from the Studio UI.
4. Given the error explained in the **Other information** section, please use: `CELERY_ALWAYS_EAGER = True` in both of your environments (LMS/CMS).
5. Go to a sections' course, click the configuration gear > Visibility > Hide in course outline
6. As an instructor, you should be able to see the hidden section with its children
7. As a student, you should be able to see the hidden subsections/units with the URL from the blocks' `View Live` option

https://github.com/openedx/edx-platform/assets/64440265/659c2bee-adbc-468a-a2ed-bf6c4ce0cbd8

## Deadline

This effort is part of the Spanish consortium project, so it'd be ideal to merge this before the end of the project.

## Other information

After deploying these changes to a tutor nightly remote installation, we found that the changes weren't reflected on the Course Outline after saving the new visibility changes for the 1st time. Now, the second time we save the changes are triggered. We don't have an explanation for it yet, but we're actively working on it. Here's a video of the behavior:

https://github.com/openedx/edx-platform/assets/64440265/3785ec67-50f6-4281-a2da-82b7c6dcf906

Update: This was solved by https://github.com/openedx/edx-platform/pull/34020, which is already under review.
